cogs/wiki.py

```diff
@@ -38,42 +44,90 @@ def __init__(self, bot: Bot):
         urls = [response['query']['pages'][str(pageid)]['fullurl'] for pageid in pageids]
         return urls
 
+    async def read_titles(self, namespace):
+        """Helper function to read all the page titles from a single namespace.
+        Sandbox link for this query:
+        https://cavesofqud.gamepedia.com/Special:ApiSandbox#action=query&format=json&prop=&list=allpages&pageids=1&apnamespace=0
+        """
+        fresh_titles = {}
+        # there's a limit on how many titles we can fetch at a time (currently 5000 for bots)
+        # and we may have more wiki articles than that someday, so fetch in batches
+        got_all = False
+        apfrom = ''  # the article title to 'continue' querying from, if necessary
+        while not got_all:
+            params = {'format': 'json',
+                      'action': 'query',
+                      'list': 'allpages',
+                      'apfrom': apfrom,
+                      'apnamespace': namespace,
+                      'aplimit': 5000}  # TODO: add to config
+            async with http_session.get(url=self.url, params=params) as reply:
+                response = await reply.json()
+            new_items = response['query']['allpages']
+            for item in new_items:
+                title = item['title']
+                # filter out some Cargo tables that shouldn't be in main namespace
+                if not title.startswith('DynamicObjectsTable:'):
+                    fresh_titles[title] = item['pageid']
+            if 'continue' in response:
+                apfrom = response['continue']['apcontinue']
+            else:
+                got_all = True
+        return fresh_titles
+
+    async def refresh_titles_cache(self):
+        """Helper function to get, or refresh, all relevant page titles for our custom search."""
+        # TODO: put cache time limit in config
+        # use time.monotonic because we don't actually care what time it is, only that it ticks,
+        # and can't go backwards due to time zone change/daylight savings.
+        if self.all_titles != {} and time.monotonic() - self.all_titles_stamp < 900:
+            # we have cached titles, and they're less than 15 minutes old
+            return
+        # else, we need to fetch new titles and update timestamp
+        new_titles = {}
+        namespaces = [0, 14]  # Main, Category  TODO: add to config
+        for namespace in namespaces:
+            new_titles.update(await self.read_titles(namespace))
+        self.all_titles = new_titles
+        self.all_titles_stamp = time.monotonic()
+
     @command()
     async def wiki(self, ctx: Context, *args):
         """Search titles of articles for the given text."""
```

Comment:
Maybe add a note about how it first searches by returns multiple pages if there are multiple hits. Does this use fuzzy search too if there are no results? Does it use it after/instead of the normal keyword extended search? I've noticed with the current cryptogull that uses extended wiki search, searching multiple keywords make some optional ex. searching `folded carbide long` will drop "long" and return the page for Two-handed folded carbide battle axe as one result. Little quirks like this should probably be noted?

Reply:
The way the old `?wiki` search (title search) worked was if the entire query was "in" an article title (like how "carbide" is in "two-handed carbide battle axe"), it would accept it as a result IF there hadn't been a full title match.

I've added a description of how the new fuzzy search works in the code.